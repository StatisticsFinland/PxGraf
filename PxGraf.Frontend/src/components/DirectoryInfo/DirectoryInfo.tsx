import { Link } from "react-router-dom";
import { ListItemText, ListItemButton, Typography, Divider } from '@mui/material';
import { urls } from 'Router';
import React from "react";
import { spacing } from 'utils/componentHelpers';
import { MultiLanguageString } from "../../types/multiLanguageString";
import { UiLanguageContext } from 'contexts/uiLanguageContext';

interface IDirectoryInfoProps {
    path: string;
    item: {
        id?: string;
        updated?: string;
        type?: 't' | 'l';
        text?: MultiLanguageString;
        languages?: string[];
    }
}

export const DirectoryInfo: React.FC<IDirectoryInfoProps> = ({ path, item }) => {
    const { language } = React.useContext(UiLanguageContext);
    const displayLanguage = item.languages.includes(language) ? language : item.languages[0];

    const currentPath = [...path, item.id];
    return (
        <>
            <ListItemButton id="mainContent" component={Link} to={urls.tableList(currentPath)}>
                <ListItemText primary={
                    <>
                        <Typography variant="body1" sx={{ ...spacing(1) }}>
                            {item.text[displayLanguage]}
                        </Typography>
                        {
                            item.updated &&
                            <Typography variant="body2" sx={{ ...spacing(1) }}>
                                {item.updated}
                            </Typography>
                        }
                    </>
                } />
            </ListItemButton>
            <Divider />
        </>
    );
}
export default DirectoryInfo;