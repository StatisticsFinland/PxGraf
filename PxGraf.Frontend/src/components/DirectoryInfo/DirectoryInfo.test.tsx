import { render, screen } from '@testing-library/react';
import { DirectoryInfo } from './DirectoryInfo';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import UiLanguageContext from 'contexts/uiLanguageContext';

const mockPath = 'asd123';
const mockItem = {
    id: 'id',
    updated: '01.01.2000',
    text: { 'fi': 'foo1-fi', 'en': 'foo1-en', 'sv': 'foo1-sv' },
    languages: ['fi', 'en', 'sv']
};

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => undefined),
            },
        };
    },
}));

jest.mock('api/services/table', () => ({
    ...jest.requireActual('api/services/table'),
    useTableQuery: () => {
        return mockItem;
    },
}));

const setLanguage = jest.fn();
const language = 'fi';

const setLanguageTab = jest.fn();
const languageTab = 'fi';

const availableUiLanguages = ['fi', 'en', 'sv'];
const uiContentLanguage = "fi";
const setUiContentLanguage = jest.fn();

describe('Rendering test', () => {
    it('renders correctly', () => {
        // mock router for testing purposes
        const { asFragment } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <DirectoryInfo path={mockPath} item={mockItem} />
                </UiLanguageContext.Provider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('should contain a specific link with a specific text', () => {
        const expectedHref = '/table-list/a/s/d/1/2/3/id/';
        render(<MemoryRouter><DirectoryInfo path={mockPath} item={mockItem} /></MemoryRouter>);
        expect(screen.getByText(mockItem.text[language])).toBeInTheDocument();
        expect(screen.getByRole('link').getAttribute('href')).toEqual(expectedHref);
    });

    it('displays text in the correct language when the language is included in item.languages', () => {
        const { getByText } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <DirectoryInfo path={mockPath} item={mockItem} />
                </UiLanguageContext.Provider>
            </MemoryRouter>
        );

        expect(getByText('foo1-fi')).toBeInTheDocument();
    });

    it('displays text in the default language when the language is not included in item.languages', () => {
        const { getByText } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language: 'fr', setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <DirectoryInfo path={mockPath} item={mockItem} />
                </UiLanguageContext.Provider>
            </MemoryRouter>
        );

        expect(getByText('foo1-fi')).toBeInTheDocument();
    });
});