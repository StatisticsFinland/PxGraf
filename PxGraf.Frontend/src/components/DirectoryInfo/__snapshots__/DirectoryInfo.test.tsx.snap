// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <a
    class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
    href="/table-list/a/s/d/1/2/3/foo-group/"
    id="mainContent"
    tabindex="0"
  >
    <div
      class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-5arffz-MuiTypography-root"
      >
        foo1-fi
      </p>
    </div>
  </a>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
  />
</DocumentFragment>
`;
